File: backend/app/services/admin_coordenador_service.rb
# frozen_string_literal: true

class AdminCoordenadorService
  def initialize(
    firestore_service: FirestoreService.new,
    firebase_auth_client: FirebaseAuthClient.new
  )
    @firestore = firestore_service
    @firebase_auth = firebase_auth_client
  end

  def listar
    @firestore.listar_usuarios_por_role("COORDENADOR").map { |doc| para_model(doc) }
  end

  def buscar_por_id(id)
    doc = @firestore.buscar_usuario_por_id(id)
    return nil unless doc
    return nil unless FirestoreKeyNormalizer.role_matches?(role_do_usuario(doc), "coordenador")

    para_model(doc)
  end

  def criar(coordenador)
    validar(coordenador)
    uid = @firebase_auth.upsert_user(
      preferred_uid: nil,
      email: coordenador[:email],
      display_name: coordenador[:nome],
      password: senha_padrao_coordenador(coordenador[:email])
    )
    @firestore.salvar_usuario(uid, payload(uid, coordenador))
    buscar_por_id(uid) || raise(ArgumentError, "Falha ao criar coordenador.")
  end

  def atualizar(id, coordenador)
    validar(coordenador)
    return nil unless buscar_por_id(id)

    @firebase_auth.upsert_user(
      preferred_uid: id,
      email: coordenador[:email],
      display_name: coordenador[:nome],
      password: senha_padrao_coordenador(coordenador[:email])
    )
    @firestore.salvar_usuario(id, payload(id, coordenador))
    buscar_por_id(id)
  end

  def remover(id)
    return false unless buscar_por_id(id)

    @firebase_auth.delete_by_uid(id)
    @firestore.remover_usuario(id)
    true
  end

  private

  def validar(coordenador)
    raise ArgumentError, "Nome e e-mail do coordenador sao obrigatorios." if coordenador.blank?
    raise ArgumentError, "Nome e e-mail do coordenador sao obrigatorios." if coordenador[:nome].blank? || coordenador[:email].blank?
  end

  def payload(uid, coordenador)
    {
      "uid" => uid,
      "nome" => coordenador[:nome].to_s.strip,
      "email" => coordenador[:email].to_s.strip.downcase,
      "role" => "coordenador",
      "departamento" => coordenador[:departamento].to_s.strip,
      "status" => coordenador[:status].presence || "ativo",
      "cursos" => coordenador[:cursos] || [],
      "vinculo" => {}
    }
  end

  def para_model(doc)
    data = doc[:data]
    cursos = FirestoreKeyNormalizer.fetch_field(data, "cursos") || []

    {
      id: doc[:id],
      nome: FirestoreKeyNormalizer.fetch_field(data, "nome"),
      email: FirestoreKeyNormalizer.fetch_field(data, "email"),
      departamento: FirestoreKeyNormalizer.fetch_field(data, "departamento"),
      status: FirestoreKeyNormalizer.fetch_field(data, "status"),
      cursos: cursos.is_a?(Array) ? cursos : []
    }
  end

  def senha_padrao_coordenador(email)
    normalized = email.to_s.strip.downcase
    at_index = normalized.index("@")
    base = at_index.to_i.positive? ? normalized[0...at_index] : normalized
    raise ArgumentError, "E-mail invalido para gerar senha padrao do coordenador." if base.blank?

    "#{base}2026"
  end

  def role_do_usuario(doc)
    FirestoreKeyNormalizer.fetch_field(doc[:data], "role", "perfil", "tipo")
  end
end

File: backend/src/services/adminCoordenadorService.js
const { IllegalArgumentError } = require('../middleware/errorHandler');
const { rolesEquivalentes } = require('../utils/roleUtils');

class AdminCoordenadorService {
  constructor(firestoreService, firebaseUserProvisioningService) {
    this.firestoreService = firestoreService;
    this.firebaseUserProvisioningService = firebaseUserProvisioningService;
  }

  async listar() {
    const docs = await this.firestoreService.listarUsuariosPorRole('COORDENADOR');
    return docs.map((d) => this.paraModel(d));
  }

  async buscarPorId(id) {
    const doc = await this.firestoreService.buscarUsuarioPorId(id);
    if (!doc) return null;
    if (!rolesEquivalentes(doc.get('role'), 'COORDENADOR')) return null;
    return this.paraModel(doc);
  }

  async criar(coordenador) {
    this.validar(coordenador);
    const uid = await this.firebaseUserProvisioningService.upsertUser(
      null,
      coordenador.email,
      coordenador.nome,
      this.senhaPadraoCoordenador(coordenador.email)
    );
    await this.firestoreService.salvarUsuario(uid, this.payload(uid, coordenador));
    return await this.buscarPorId(uid);
  }

  async atualizar(id, coordenador) {
    this.validar(coordenador);
    const atual = await this.buscarPorId(id);
    if (!atual) return null;

    await this.firebaseUserProvisioningService.upsertUser(
      id,
      coordenador.email,
      coordenador.nome,
      this.senhaPadraoCoordenador(coordenador.email)
    );

    await this.firestoreService.salvarUsuario(id, this.payload(id, coordenador));
    return await this.buscarPorId(id);
  }

  async remover(id) {
    const atual = await this.buscarPorId(id);
    if (!atual) return false;
    await this.firebaseUserProvisioningService.deleteByUid(id);
    await this.firestoreService.removerUsuario(id);
    return true;
  }

  validar(coordenador) {
    if (!coordenador || this.isBlank(coordenador.nome) || this.isBlank(coordenador.email)) {
      throw new IllegalArgumentError('Nome e e-mail do coordenador sao obrigatorios.');
    }
  }

  payload(uid, c) {
    return {
      uid,
      nome: String(c.nome).trim(),
      email: String(c.email).trim().toLowerCase(),
      role: 'COORDENADOR',
      departamento: c.departamento == null ? '' : String(c.departamento).trim(),
      status: c.status == null ? 'Ativo' : String(c.status).trim(),
      cursos: c.cursos == null ? [] : c.cursos,
      vinculo: {}
    };
  }

  paraModel(doc) {
    return {
      id: doc.id,
      nome: this.texto(doc.get('nome')),
      email: this.texto(doc.get('email')),
      departamento: this.texto(doc.get('departamento')),
      status: this.texto(doc.get('status')),
      cursos: doc.get('cursos') || []
    };
  }

  senhaPadraoCoordenador(email) {
    const normalized = email == null ? '' : String(email).trim().toLowerCase();
    const atIndex = normalized.indexOf('@');
    const base = atIndex > 0 ? normalized.substring(0, atIndex) : normalized;
    if (!base.trim()) throw new IllegalArgumentError('E-mail invalido para gerar senha padrao do coordenador.');
    return base + '2026';
  }

  isBlank(v) {
    return v == null || String(v).trim().length === 0;
  }

  texto(v) {
    return v == null ? '' : String(v);
  }
}

module.exports = AdminCoordenadorService;

